graphix/generator.py

```diff
@@ -68,55 +70,120 @@ def generate_from_graph(
-                pattern.add(X(node=f[j].pop(), domain={j}))
-    else:
-        # no flow found - we try gflow
-        g, l_k = find_gflow(graph, set(inputs), set(outputs), meas_planes=meas_planes)
-        if g is not None:
-            # gflow found
-            depth, layers = get_layers(l_k)
-            pattern = Pattern(input_nodes=inputs)
-            for i in set(graph.nodes) - set(inputs):
-                pattern.add(N(node=i))
-            for e in graph.edges:
-                pattern.add(E(nodes=e))
-            for i in range(depth, 0, -1):  # i from depth, depth-1, ... 1
-                for j in layers[i]:
-                    pattern.add(M(node=j, plane=meas_planes[j], angle=angles[j]))
-                    odd_neighbors = find_odd_neighbor(graph, g[j])
-                    for k in odd_neighbors - {j}:
-                        pattern.add(Z(node=k, domain={j}))
-                    for k in g[j] - {j}:
-                        pattern.add(X(node=k, domain={j}))
-        else:
-            raise ValueError("no flow or gflow found")
-
-    pattern.reorder_output_nodes(outputs)
+        pattern = _flow2pattern(graph, angles, inputs, f, l_k)
+        pattern.reorder_output_nodes(outputs)
+        return pattern
+
+    # no flow found - we try gflow
+    g, l_k = find_gflow(graph, inputs, outputs, meas_planes=meas_planes)
+    if g is not None:
+        # gflow found
+        pattern = _gflow2pattern(graph, angles, inputs, meas_planes, g, l_k)
+        pattern.reorder_output_nodes(outputs)
+        return pattern
+
+    # no flow or gflow found - we try pflow
+    p, l_k = find_pauliflow(graph, inputs, outputs, meas_planes=meas_planes, meas_angles=angles)
+    if p is not None:
+        # pflow found
+        pattern = _pflow2pattern(graph, angles, inputs, meas_planes, p, l_k)
+        pattern.reorder_output_nodes(outputs)
+        return pattern
+
+    raise ValueError("no flow or gflow or pflow found")
+
+
+def _flow2pattern(
+    graph: nx.Graph[int],
+    angles: Mapping[int, ExpressionOrFloat],
+    inputs: Iterable[int],
+    f: Mapping[int, AbstractSet[int]],
+    l_k: Mapping[int, int],
+) -> Pattern:
+    """Construct a measurement pattern from a causal flow according to the theorem 1 of [NJP 9, 250 (2007)]."""
+    depth, layers = get_layers(l_k)
+    pattern = Pattern(input_nodes=inputs)
+    for i in set(graph.nodes) - set(inputs):
+        pattern.add(N(node=i))
+    for e in graph.edges:
+        pattern.add(E(nodes=e))
+    measured = []
+    for i in range(depth, 0, -1):  # i from depth, depth-1, ... 1
+        for j in layers[i]:
+            measured.append(j)
+            pattern.add(M(node=j, angle=angles[j]))
+            neighbors: set[int] = set()
+            for k in f[j]:
+                neighbors |= set(graph.neighbors(k))
+            for k in neighbors - {j}:
+                # if k not in measured:
+                pattern.add(Z(node=k, domain={j}))
+            fj, *_ = f[j]
```

Comment:
We can make it more readable the following way:
```suggestion
            (fj, ) = f[j]
```
since for a causal flow `f[j]` is a set with always one element.